java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java

```diff
@@ -972,26 +974,168 @@ public Description matchVariable(VariableTree tree, VisitorState state) {
+    COMMA(","),
+    SEMICOLON(";"),
+    NONE("");
+    @Nullable final String endingChar;
+
+    @Nullable
+    String getChar() {
+      return endingChar;
+    }
+
+    EnumEnding() {
+      this.endingChar = null;
+    }
+
+    EnumEnding(@Nullable final String endingChar) {
+      this.endingChar = endingChar;
+    }
+  }
+
+  private Stream<VariableTree> getEnumConstants(ClassTree enumClassTree) {
+    return enumClassTree
+        .getMembers()
+        .stream()
+        .filter(member -> member.getKind() == Kind.VARIABLE)
+        .map(member -> ((VariableTree) member))
+        .filter(member -> ASTHelpers.getSymbol(member).getKind() == ElementKind.ENUM_CONSTANT);
+  }
+
+  private boolean isOnlyEnumConstant(Symbol enumSym, VisitorState state) {
+    ClassTree enumClassTree = ASTHelpers.findClass(enumSym.enclClass(), state);
+
+    if (enumClassTree == null) {
+      return false;
+    }
+
+    // True if there is only a single enum constant for this enum
+    return getEnumConstants(enumClassTree).count() == 1;
+  }
+
+  /**
+   * This "looks ahead" at the next enum constant in an enum definition to see if it will be cleaned
+   * up by Piranha. That way, we can know what kind of character to append to the end of the
+   * previous enum constant after the last one is removed. This is important for enums that have non
+   * enum constant members after their enum constants, such as constructors, methods, and fields.
+   */
+  private EnumEnding getEndingOfLastEnumConstantIfRemoved(Symbol enumSym, VisitorState state) {
+    ClassTree enumClassTree = ASTHelpers.findClass(enumSym.enclClass(), state);
+
+    if (enumClassTree == null) {
+      return EnumEnding.IGNORE;
+    }
+
+    // Find enum initializer members only
+    List<VariableTree> enumConstants = getEnumConstants(enumClassTree).collect(Collectors.toList());
+
+    // Get second to last index
+    //
+    // The reason to get the second to last index, is that we need to alter the second to last enum
+    // constant ending if the last constant is being removed, as the second to last enum constant
+    // will now become the last constant.
+    //
+    // In Java, the last enum constant is be required to have a different ending than other enum
+    // constants in the enum if the enum has non-enum constant members
+    //
+    // We do not need to alter any other enum constants because they will either never become the
+    // last constant [0 ... size - 3] or will be removed [size - 1].
+    int index = enumConstants.size() - 2;
+
+    // Check if the second to last enum is the current enum being processed
+    if (index < 0 || !enumConstants.get(index).getName().equals(enumSym.getSimpleName())) {
+      return EnumEnding.IGNORE;
+    }
+
+    VariableTree nextConstant = enumConstants.get(index + 1);
+    String nextEnumConstantSource = state.getSourceForNode(nextConstant);
+    String nextEnumSource = state.getSourceForNode(state.getPath().getParentPath().getLeaf());
```

Comment:
What does `nextEnumSource` mean here? Is this the source for the entirety of the enclosing `enum Foo {...}` tree?

Reply:
Yes! Maybe a better name is `enumSourceForCurrentAndNextConstant`? As it should always be the source for the enclosing enum type that contains both the current and next enum constant.